shell/renderer/api/atom_api_web_frame.cc

```diff
@@ -436,6 +477,102 @@ v8::Local<v8::Promise> ExecuteJavaScriptInIsolatedWorld(
   return handle;
 }
 
+v8::Local<v8::Value> ExecuteJavaScriptSync(gin_helper::Arguments* args,
+                                           v8::Local<v8::Value> window,
+                                           const base::string16& code) {
+  v8::Isolate* isolate = args->isolate();
+  bool has_user_gesture = false;
+  args->GetNext(&has_user_gesture);
+
+  // return v8::Null(isolate);
+
+  ScriptSyncExecutionCallback* sync_callback =
+      new ScriptSyncExecutionCallback(isolate);
+
+  GetRenderFrame(window)->GetWebFrame()->RequestExecuteScriptAndReturnValue(
+      blink::WebScriptSource(blink::WebString::FromUTF16(code)),
+      has_user_gesture, sync_callback);
+
+  if (sync_callback->completed_) {
+    v8::Local<v8::Value> result =
+        v8::Local<v8::Value>::New(isolate, sync_callback->result_);
+    std::string error_message = sync_callback->error_message_;
+    delete sync_callback;
+    if (error_message.empty()) {
+      return result;
+    } else {
+      args->ThrowError(error_message);
+      return v8::Null(isolate);
+    }
+  } else {
+    sync_callback->delete_self_ = true;
+
+    // this can only happen if the frame is not in the
+    // mojom::FrameLifecycleState::kRunning state
+    args->ThrowError("execution did not complete synchronously");
```

Comment:
This makes no sense as an API, their is no way the app can handle this in any useful way.

Reply:
I have never seen the top frame not in the mojom::FrameLifecycleState::kRunning state.

Can you point out when it wouldn't be?

When a frame is in the kRunning state, execution is guaranteed to be synchronous, that's not an "implementation detail" it's a public api guarantee

```c++
  enum ScriptExecutionType {
    // Execute script synchronously, unless the page is suspended.
    kSynchronous,
    // Execute script asynchronously.
    kAsynchronous,
    // Execute script asynchronously, blocking the window.onload event.
    kAsynchronousBlockingOnload
  };
```

Only if the page is suspended, would execution not be synchronous.

If the top frame being suspended is an unavoidable, unpredictable possibility outside of one's control, which seems unlikely, I would agree with you. Otherwise these sync methods are useful and will work with 100% reliability and so should not be rejected.

Reply:
I believe this could happen if the window is displaying an `alert` or similar: https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/page/page.h?l=237-244&rcl=879cf27d2d56f2ff63e2ba7041bfabdd2d96deb3

Reply:
I'll test it. If so, and if this is something that could be detected, the sync methods could fail early. That would not be unreasonable behavior.

Reply:
I'd also test what happens if the executed script calls `alert`; that seems like it would pause the script also and prevent a synchronous return from happening.